### PR TITLE
Agentic Commerce: reserve stock at order completion so concurrent sessions cannot oversell

### PR DESCRIPTION
**Pull request by wjrosa**
Fixes STRIPE-1256

## Changes proposed in this Pull Request:

The agentic checkout flow validated stock at `finalize_checkout` without holding it, then created and completed the order after Stripe had already captured the payment, with no re-check. Concurrent sessions could each pass validation for the same last units and all complete — driving stock negative with paid but unfulfillable orders.

- **Atomic hold at order creation:** the order mapper now places a WC `ReserveStock` hold before `payment_complete()`, so concurrent completions serialize on the `wc_reserved_stock` table. The hold duration is passed explicitly so a blank `woocommerce_hold_stock_minutes` can't disable the guard; `payment_complete()`'s stock reduction releases it moments later.
- **Explicit oversell handling:** if the hold fails, the already-paid order is parked **on-hold** with the transaction id and an explanatory note. Stock reduction is suppressed via `woocommerce_payment_complete_reduce_order_stock` (not `woocommerce_can_reduce_order_stock`, which would still flag the order stock-reduced and make a later cancel/refund restock units never taken).
- **Backorders honored:** `ReserveStock` skips backorder-enabled products, and finalize validation no longer declines quantities backorders permit. Finalize also subtracts `wc_get_held_stock_quantity()` so it stops approving units concurrent checkouts already hold.

**BC impact:** no public signatures, hooks, or options change. An out-of-stock-at-completion agentic order now ends `on-hold` instead of `processing` with negative stock; `ReserveStock` exists throughout the supported WC range.

## Testing instructions

1. With the agentic feature flag on, set a product to managed stock, quantity 1.
2. Run two overlapping agentic checkouts for it (or insert a `wc_reserved_stock` row for another order to simulate the second). One order completes as `processing` with stock 0; the other is created `on-hold` with a note, stock never goes to −1, and the payment intent is on the order for refunding.
3. With backorders set to "Allow", both complete and stock may go negative.
4. `npm run test:php -- --filter "WC_Stripe_Agentic_Commerce_Order_Mapper_Test|WC_Stripe_Agentic_Commerce_Manual_Approval_Test"` includes a regression test for the concurrent race.

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

Changelog entries added to `changelog.txt` and `readme.txt` under 10.9.0.

<details>

<summary>Changelog Entry Comment</summary>

#### Comment

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: includes/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper.php [hunk 1]
```diff
@@ -73,6 +84,16 @@ public function create_order_from_checkout_session( WC_Stripe_Agentic_Checkout_S
 			throw $e;
 		}
 
+		// The only point that serializes concurrent agentic sessions on stock —
+		// finalize_checkout validates without reserving, so N sessions can all
+		// pass for the same last units. ReserveStock either claims them or throws.
+		try {
+			$this->reserve_stock( $order );
+		} catch ( \Automattic\WooCommerce\Checkout\Helpers\ReserveStockException $e ) {
```
Comment (jaclync) [inline includes/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper.php]: ❓ What happens when other types of exceptions are throw from `reserve_stock`, like some unexpected errors from third-party plugins in the callbacks in `reserve_stock_for_order`? Since the catch here just handles `ReserveStockException`, other exception types might return after the try, skipping both `payment_complete` and the on-hold fallback. This can leave a charged customer pending. Core only throws `ReserveStockException`, it's likely an edge case.

File: tests/phpunit/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper-test.php [hunk 2]
```diff
@@ -1571,6 +1573,152 @@ public function test_exception_thrown_when_order_total_mismatches_session() {
...
+		$competing = wc_create_order( [ 'status' => OrderStatus::PENDING ] );
+		$competing->add_product( wc_get_product( $product->get_id() ), 1 );
+		$competing->save();
+		( new \Automattic\WooCommerce\Checkout\Helpers\ReserveStock() )->reserve_stock_for_order( $competing, 10 );
+
+		$session = $this->build_stock_session( $product, 1 );
+		$order   = $this->mapper->create_order_from_checkout_session( $session );
+
+		$this->assertEquals( OrderStatus::ON_HOLD, $order->get_status() );
+		$this->assertSame( 1, wc_get_product( $product->get_id() )->get_stock_quantity(), 'Stock must not be reduced for the oversold order.' );
```
Comment (jaclync) [inline tests/phpunit/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper-test.php:1614]: nit: since `woocommerce_can_reduce_order_stock` filter was specifically not chosen so as not to restock after a later cancel/refund, could this test also assert on this behavior?

File: includes/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper.php [hunk 3]
```diff
@@ -294,6 +315,65 @@ private function add_product_to_order( WC_Order $order, WC_Product $product, int
...
+			return (int) $target_order_id === $order_id ? false : $trigger_reduce;
+		};
+		add_filter( 'woocommerce_payment_complete_reduce_order_stock', $block_reduction, 10, 2 );
+
+		try {
+			$order->update_status(
+				OrderStatus::ON_HOLD,
+				sprintf(
+					/* translators: %s: reason stock could not be secured */
+					__( 'Stripe captured the payment, but stock could not be secured for every item: %s Review stock, then process or refund this order manually.', 'woocommerce-gateway-stripe' ),
```
Comment (jaclync) [inline includes/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper.php]: nit question: do all reasons end with a period or some separator in all supported languages? If not, the sentence might be connected to the next (`Review stock...`), just wondering. An alternative is to have the freeform reason at the end.

**Review by jaclync — APPROVED** [review]
LGTM! How do you usually test the agentic checkout flow? With ChatGPT/Claude chats set up in some way?

Tested locally with AI-assisted scripts, with WC 10.9.4: happy/oversold paths, and two parallel PHP processes into the mapper for the same remaining quantity with a shared start barrier - one order in processing, the other became on-hold ✅

File: includes/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper.php (hunk 1 above)
Comment (wjrosa) [reply includes/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper.php]: Good catch — now catches Throwable so any reservation failure parks the order; test added with a throwing third-party callback (b0bae7ae7).

> *Drafted with AI; reviewed by me.*

File: tests/phpunit/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper-test.php (hunk 2 above)
Comment (wjrosa) [reply tests/phpunit/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper-test.php:1614]: Added — the race test now cancels the parked order and asserts stock stays untouched (b0bae7ae7).

> *Drafted with AI; reviewed by me.*

File: includes/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper.php (hunk 3 above)
Comment (wjrosa) [reply includes/agentic-commerce/class-wc-stripe-agentic-commerce-order-mapper.php]: Moved the freeform reason to the end ("… Reason: %s") in b0bae7ae7.

> *Drafted with AI; reviewed by me.*

**Comment by wjrosa** [comment]
Thanks for the review, @jaclync!

> How do you usually test the agentic checkout flow? With ChatGPT/Claude chats set up in some way?

It is a bit tricky to test it. I have collected a better guide using my other PRs as references:

- Connect Stripe in test mode.
- Enable the developer feature flag: `wp option update _wcstripe_feature_agentic_commerce yes` (or return true from the `wc_stripe_is_agentic_commerce_enabled` filter).
- Enable the merchant toggle in WooCommerce → Settings → Payments → Stripe → Agentic Commerce.
- Follow the instructions there to perform the onboarding on the Stripe Dashboard side
- Trigger an initial catalog sync
- Go to the Stripe Dashboard, navigate to `/test/agentic-commerce`
- Enable the *Test Agent*
- Click the "View Feed" button at the top card
- In the product list, click "Test"

This will open a new screen that allows you to simulate the entire agentic purchase flow, as if you were doing it using real agents. Currently, there's no way to perform tests using real agents

Screenshot 2026-07-23 at 10 33 16

Feel free to reach me if you have any questions/issues!

**Comment by jaclync** [comment]
> * Follow the instructions there to perform the onboarding on the Stripe Dashboard side

👋 Does the Stripe account need to be enabled for some private preview feature on the Stripe end? I reached this step, but couldn't see the `Agentic commerce` menu under Payments following this step in wp-admin:

> Log into your [Stripe Dashboard↗](https://dashboard.stripe.com/test/agentic-commerce) and go to Payments > Agentic commerce

No worries if this feature isn't what everyone on the team needs for end-to-end testing.

**Comment by wjrosa** [comment]
You should be able to onboard yourself by now. Asked the Stripe team (and tagged you) in p1785264306292189-slack-C9976E5MJ
